### PR TITLE
feat: alluxio-py support alluxio and oss filesystem

### PR DESCRIPTION
**Pull request by liiuzq-xiaobai**
The implementation of the delegated filesystem for Alluxio and OSS has been completed. Specific notes:
	1.Users need to specify in the configuration whether the delegated filesystem should be accelerated by Alluxio using the alluxio_enable flag. If set to true, the configuration file must still include the necessary initialization settings for the Alluxio filesystem.
	2.The configuration file can include multiple OSS filesystems as delegated filesystems, but it is necessary to ensure that their bucket_name is unique. A unique delegated filesystem is determined by the combination of the delegated filesystem name and the bucket_name.

### CODE REVIEW

File: alluxio/posix/config.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    def __init__(self, config_file_path='config/ufs_config.yaml'):
+        self.config_file_path = config_file_path
+        self.config_data = self._load_config()
+        self.validation_functions = {
+            Constants.OSS_FILESYSTEM_TYPE: self._validate_oss_config,
+            Constants.ALLUXIO_FILESYSTEM_TYPE: self._validate_alluxio_config,
+            Constants.S3_FILESYSTEM_TYPE: self._validate_s3_config
+        }
+
+    def _load_config(self):
```
Comment (ccy00808) [inline alluxio/posix/config.py:18]: It is necessary to add a hot update interface. For example, if the user does not have a configuration file and wants to configure config through the API

File: alluxio/posix/config.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        except KeyError:
+            raise ConfigMissingError(fs_name, "FileSystem Configuration is missing")
+        except ValueError as e:
+            raise ConfigMissingError(fs_name, str(e))
+
+    def get_config_fs_list(self) -> list:
+        return self.config_data.keys()
+
+    @staticmethod
+    def _validate_oss_config(config):
```
Comment (ccy00808) [inline alluxio/posix/config.py:47]: File separation for each ufs check? A lot of ufs will definitely be expanded in the future.

File: alluxio/posix/config.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        else:
+            # If ALLUXIO_ETCD_ENABLE is False, ALLUXIO_WORKER_HOSTS must be set
+            required_keys.append(Constants.ALLUXIO_WORKER_HOSTS)
+
+        if not all(config.get(key) for key in required_keys):
+            raise ConfigMissingError(f"The following keys must be set in the configuration: {required_keys}")
+
+
+    @staticmethod
+    def _validate_s3_config(config):
```
Comment (ccy00808) [inline alluxio/posix/config.py:76]: The s3 interface also needs to be implemented

File: alluxio/posix/delegate.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import os
+from alluxio.posix import fileimpl
+config_manager = fileimpl.ConfigManager("../../config/ufs_config.yaml")
```
Comment (ccy00808) [inline alluxio/posix/delegate.py:3]: The config path cannot be hard-coded in the code. It must have the feasibility of dynamic configuration, such as environment variables and API injection.

File: alluxio/posix/demo.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
+import os
+import delegate
+
+
+def delegatefs_open_write():
+    write_file_path = f'oss://alhz-ossp-alluxio-test/alluxio-py/delegatefs-io-1.txt'
```
Comment (ccy00808) [inline alluxio/posix/demo.py:6]: please move this file to tests.

File: alluxio/posix/config.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            return fs_config
+        except KeyError:
+            raise ConfigMissingError(fs_name, "FileSystem Configuration is missing")
+        except ValueError as e:
+            raise ConfigMissingError(fs_name, str(e))
+
+    def get_config_fs_list(self) -> list:
+        return self.config_data.keys()
+
+    def update_config(self, fs_type, key, value):
```
Comment (ccy00808) [inline alluxio/posix/config.py:56]: Pass in key value through **

File: alluxio/posix/config.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        self.config_file_path = os.getenv('ALLUXIO_PY_CONFIG_FILE_PATH', 'config/ufs_config.yaml')
+        self.config_data = self._load_config()
+        self.validation_functions = {
+            Constants.OSS_FILESYSTEM_TYPE: validate_oss_config,
+            Constants.ALLUXIO_FILESYSTEM_TYPE: validate_alluxio_config
+        }
+
+    def _load_config(self):
+        if not os.path.exists(self.config_file_path):
+            raise FileNotFoundError(f"{self.config_file_path} does not exist.")
```
Comment (ccy00808) [inline alluxio/posix/config.py:25]: It does not necessarily depend entirely on the configuration file, and can be configured directly through hot update without a configuration file. If there is no corresponding configuration during use, an exception can be thrown directly.

File: alluxio/posix/fileimpl.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+local_listdir = os.listdir
+local_rename = os.rename
+local_close = os.close
+local_mkdir = os.mkdir
+local_remove = os.remove
+local_rmdir = os.rmdir
+
+
+def open(file: str, mode: str = "r", **kw):
+    logging.info("DelegateFileSystem opening file: %s", file)
```
Comment (ccy00808) [inline alluxio/posix/fileimpl.py:22]: info change to debug.

File: alluxio/posix/fileimpl.py [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    instance = DelegateFileSystem.instance
+    fs = instance.get_file_system(file)
+    if fs:
+        try:
+            return fs.open(file, mode, **kw)
```
Comment (ccy00808) [inline alluxio/posix/fileimpl.py:27]: Through fspec reflection, parameters can be passed directly into ufs client. Is there no need to verify the legality of config parameters?

File: alluxio/posix/fileimpl.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+            return local_open(file, mode, **kw)
+    return local_open(file, mode, **kw)
+
+
+def stat(path: str, **kw):
+    instance = DelegateFileSystem.instance
+    fs = instance.get_file_system(path)
+    if fs:
+        try:
+            logging.info("DelegateFileSystem getStatus filemeta: %s", path)
```
Comment (ccy00808) [inline alluxio/posix/fileimpl.py:41]: the log level

File: alluxio/posix/fileimpl.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+        except Exception as e:
+            logging.error(
+                f"Failed to rename file by delegateFileSystem with exception: {e}."
+                f"Used local filesystem instead.")
+            return local_rename(src, dest, **kw)
+    logging.error("Source and destination are on different file systems or not supported.")
+    return local_rename(src, dest, **kw)
+
+
+class DelegateFileSystem:
```
Comment (ccy00808) [inline alluxio/posix/fileimpl.py:125]: Separate independent files

File: alluxio/posix/fileimpl.py [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+        self.config_manager = config_manager
+        self.filesystem_storage = FSStorage()
+        self.filesystem_storage.data = {}
+        self.enableFileSystems = [Constants.OSS_FILESYSTEM_TYPE,
+                                  Constants.ALLUXIO_FILESYSTEM_TYPE,
+                                  Constants.S3_FILESYSTEM_TYPE]
+        self.__init__file__system()
+        DelegateFileSystem.instance = self
+
+    def __create__file__system(self, fs_name: str):
```
Comment (ccy00808) [inline alluxio/posix/fileimpl.py:138]: Is it impossible to distinguish uniqueness only through fs name? For example, there are multiple buckets under the same ufs.

**Review by ccy00808 — APPROVED** [review]

**Review by ccy00808 — APPROVED** [review]

**Comment by YichuanSun** [comment]
alluxio-bot, merge this please.

**Comment by ccy00808** [comment]
alluxio-bot, merge this please.

**Pull request merged**